src/components/DropMenu/DropMenu.jsx

```diff
@@ -0,0 +1,467 @@
+			nullOptions
+		} = this.state;
+
+		this.setState({
+			isMouseTriggered: false
+		});
+
+		if (isExpanded) {
+			if (_.includes([KEYCODE.Enter, KEYCODE.Space], e.keyCode)) {
+				e.preventDefault();
+				const focusedOptionData = _.get(flattenedOptionsData, focusedIndex, null);
+				const focusedOptionProps = _.get(focusedOptionData, 'optionProps', {});
+				if (focusedOptionData && !focusedOptionProps.isDisabled) {
+					onSelect(focusedIndex, {props: focusedOptionProps, event: e});
+				} else if (_.isNull(focusedIndex)) {
+					onSelect(null, {props: _.first(nullOptions), event: e});
+				}
+			}
+			if (e.keyCode === KEYCODE.Escape) {
+				e.preventDefault();
+				onCollapse(e);
+			}
+			if (e.keyCode === KEYCODE.ArrowUp) {
+				if (_.isNumber(focusedIndex) || _.isNull(focusedIndex)) {
+					if (focusedIndex === 0) {
+						if (!_.isEmpty(nullOptions)) {
+							e.preventDefault();
+							onFocusPrev(e);
+						}
+					}
+					if (focusedIndex > 0) {
+						e.preventDefault();
+						onFocusPrev(e);
+					}
+				} else {
+					e.preventDefault();
+					onFocusPrev(e);
+				}
+			}
+			if (e.keyCode === KEYCODE.ArrowDown) {
+				if (_.isNumber(focusedIndex)) {
+					if (focusedIndex < _.size(flattenedOptionsData) - 1) {
+						e.preventDefault();
+						onFocusNext(e);
+					}
+				} else {
+					e.preventDefault();
+					onFocusNext(e);
+				}
+			}
+		} else {
+			if (_.includes([KEYCODE.Space, KEYCODE.ArrowDown], e.keyCode)) {
+				e.preventDefault();
+				onExpand(e);
+			}
+		}
+	},
+
+	handleClick(e) {
+		const {
+			isExpanded,
+			onExpand,
+			onCollapse
+		} = this.props;
+
+		if (isExpanded) {
+			onCollapse(e);
+		} else {
+			onExpand(e);
+		}
+	},
+
+	handleChangeBounds(type) {
+		const {
+			onBelowFold,
+			onAboveFold
+		} = this.props;
+		if (type === ContextMenu.BELOW_FOLD) {
+			onBelowFold();
+		} else if (type === ContextMenu.ABOVE_FOLD) {
+			onAboveFold();
+		}
+	},
+
+	handleMouseFocusOption(optionIndex, optionProps) {
+		const {
+			focusedIndex,
+			onFocusOption
+		} = this.props;
+
+		this.setState({
+			isMouseTriggered: true
+		});
+
+		if (!optionProps.isDisabled && focusedIndex !== optionIndex) {
+			onFocusOption(optionIndex);
+		}
+	},
+
+	handleSelectOption(optionIndex, optionProps, event) {
+		const {
+			onSelect
+		} = this.props;
+
+		if (!optionProps.isDisabled) {
+			onSelect(optionIndex, { props: optionProps, event: event });
+		}
+	},
+
+	renderOption(optionProps, optionIndex, isGrouped) {
+		const {
+			selectedIndices,
+			focusedIndex,
+		} = this.props;
+
+		const {
+			isMouseTriggered
+		} = this.state;
+
+		const {
+			isDisabled
+		} = optionProps;
+
+		const isFocused = optionIndex === focusedIndex;
+		const isSelected = _.includes(selectedIndices, optionIndex);
+
+		return (
+			<div
+				key={'DropMenuOption' + optionIndex}
+				onMouseMove={() => this.handleMouseFocusOption(optionIndex, optionProps)}
+				onClick={(event) => this.handleSelectOption(optionIndex, optionProps, event)}
```

Comment:
We updated our event handling approach to help avoid introducing new functions on each render... should we be avoiding that ourselves? :)